superset/assets/src/SqlLab/components/TabbedSqlEditors.jsx

```diff
@@ -68,6 +69,20 @@ class TabbedSqlEditors extends React.PureComponent {
     this.removeAllOtherQueryEditors = this.removeAllOtherQueryEditors.bind(this);
   }
   componentDidMount() {
+    // migrate query editor and associated tables state to server
```

Comment:
do we need to run query editor migration at every time component mounted? 
i thought only need once after we enable the feature.

Reply:
@graceguo-supercat the migration is not atomic. If something fails, it will be retried later when the user opens SQL Lab again, and the state can co-exist partially in the backend and partially in `localStorage`.

Reply:
I don't understand this part...
in what cases migration will fail? and from what point sql lab will stop re-try?
after enable this feature, does sql lab still depend on localStorage?